ewms_pilot/pilot.py

```diff
@@ -246,110 +353,143 @@ async def consume_and_reply(
         raise
 
 
+async def _ack_nack_finished_tasks(
+    sub: mq.queue.ManualQueueSubResource,
+    tasks: AsyncioTaskMessages,
+    return_when: str,
+    previous_failed: AsyncioTaskMessages,
+) -> Tuple[AsyncioTaskMessages, AsyncioTaskMessages]:
+    """Get finished tasks and ack/nack their messages.
+
+    Returns:
+        Tuple:
+            AsyncioTaskMessages: pending tasks and
+            AsyncioTaskMessages: failed tasks (plus those in `previous_failed`)
+    """
+    done, pending = await asyncio.wait(tasks.keys(), return_when=return_when)
+    LOGGER.info(f"{len(done)} Tasks Finished")
+
+    for task in done:
+        if task.exception():
+            await sub.nack(tasks[task])
+            previous_failed[task] = tasks[task]
+            LOGGER.error("Task failed:")
+            LOGGER.error(_task_exception_str(task))
+        else:
+            await sub.ack(tasks[task])
```

Comment:
Is there a reason why you use `task.exception()` instead of:
```python
try:
  await task
except Exception:
  LOGGER.error("Task failed:")
````
(note that this is also how you're supposed to get the return value from the future, if you had one)

Reply:
I believe I was following examples in the docs, which used the helper methods

Reply:
I'd be interested to know which docs?